alacritty_terminal/src/term/cell.rs

```diff
@@ -94,25 +96,70 @@ impl Cell {
 
     /// Write a new zerowidth character to this cell.
     #[inline]
-    pub fn push_zerowidth(&mut self, c: char) {
-        self.extra.get_or_insert_with(Default::default).zerowidth.push(c);
-    }
+    pub(crate) fn push_zerowidth(&mut self, character: char) {
+        self.clone_or_init_extra();
 
-    /// Free all dynamically allocated cell storage.
-    #[inline]
-    pub fn drop_extra(&mut self) {
-        if self.extra.is_some() {
-            self.extra = None;
-        }
+        let extra = Rc::get_mut(self.extra.as_mut().unwrap()).unwrap();
+        extra.zerowidth.push(character);
     }
 
     /// Remove all wide char data from a cell.
     #[inline(never)]
-    pub fn clear_wide(&mut self) {
+    pub(crate) fn clear_wide(&mut self) {
         self.flags.remove(Flags::WIDE_CHAR);
-        self.drop_extra();
+        self.extra = None;
         self.c = ' ';
     }
+
+    /// Set underline color on the cell.
+    pub(crate) fn set_underline_color(&mut self, color: Option<Color>) {
+        // If we reset color and we don't have zerowidth we should drop extra storage.
+        if color.is_none()
+            && self.extra.as_ref().map(|extra| !extra.zerowidth.is_empty()) != Some(false)
+        {
+            self.extra = None;
+            return;
+        }
+
+        self.clone_or_init_extra();
+
+        let extra = Rc::get_mut(self.extra.as_mut().unwrap()).unwrap();
+        extra.underline_color = color;
+    }
+
+    /// This function initializes extra storage if it's not present, or clones the existing storage
+    /// if it's not uniquely referenced.
+    #[inline]
+    fn clone_or_init_extra(&mut self) {
+        match self.extra.as_mut() {
+            Some(extra) if Rc::strong_count(extra) == 1 => (),
+            extra => {
+                self.extra =
+                    Some(Rc::new(extra.map(|extra| (**extra).clone()).unwrap_or_default()));
+            },
+        }
+    }
+
+    /// Underline color stored in this cell.
+    #[inline]
+    pub fn undreline_color(&self) -> Option<Color> {
+        self.extra.as_ref()?.underline_color
```

Comment:
We don't return `fg`, since `fg` is being altered in renderer due to cell flags, like inversion.